tiledb/api/src/query/conditions.rs

```diff
@@ -0,0 +1,754 @@
+        match self {
+            Self::Cond(cond) => cond.build(ctx),
+            Self::Comb { lhs, rhs, op } => {
+                let lhs = lhs.build(ctx)?;
+                let rhs = rhs.build(ctx)?;
+
+                let mut c_cond: *mut ffi::tiledb_query_condition_t = out_ptr!();
+                ctx.capi_return(unsafe {
+                    ffi::tiledb_query_condition_combine(
+                        ctx.capi(),
+                        lhs.capi(),
+                        rhs.capi(),
+                        op.capi_enum(),
+                        &mut c_cond,
+                    )
+                })?;
+
+                Ok(QueryCondition {
+                    context: ctx,
+                    raw: RawQueryCondition::Owned(c_cond),
+                })
+            }
+            Self::Negate(expr) => {
+                let mut c_cond: *mut ffi::tiledb_query_condition_t = out_ptr!();
+                ctx.capi_return(unsafe {
+                    ffi::tiledb_query_condition_negate(
+                        ctx.capi(),
+                        expr.build(ctx)?.capi(),
+                        &mut c_cond,
+                    )
+                })?;
+
+                Ok(QueryCondition {
+                    context: ctx,
+                    raw: RawQueryCondition::Owned(c_cond),
+                })
+            }
+        }
+    }
+}
+
+impl BitAnd for QC {
+    type Output = Self;
+    fn bitand(self, rhs: Self) -> Self::Output {
+        QC::Comb {
+            lhs: Box::new(self),
+            rhs: Box::new(rhs),
+            op: CombineOp::And,
+        }
+    }
+}
+
+impl BitOr for QC {
+    type Output = Self;
+    fn bitor(self, rhs: Self) -> Self::Output {
+        QC::Comb {
+            lhs: Box::new(self),
+            rhs: Box::new(rhs),
+            op: CombineOp::Or,
+        }
+    }
+}
+
+impl Not for QC {
+    type Output = Self;
+    fn not(self) -> Self::Output {
+        QC::Negate(Box::new(self))
+    }
+}
+
+pub(crate) enum RawQueryCondition {
+    Owned(*mut ffi::tiledb_query_condition_t),
+}
+
+impl Deref for RawQueryCondition {
+    type Target = *mut ffi::tiledb_query_condition_t;
+    fn deref(&self) -> &Self::Target {
+        let RawQueryCondition::Owned(ref ffi) = *self;
+        ffi
+    }
+}
+
+impl Drop for RawQueryCondition {
+    fn drop(&mut self) {
+        let RawQueryCondition::Owned(ref mut ffi) = *self;
+        unsafe {
+            ffi::tiledb_query_condition_free(ffi);
+        }
+    }
+}
+
+#[derive(ContextBound)]
+pub struct QueryCondition<'ctx> {
+    #[context]
+    context: &'ctx Context,
+    raw: RawQueryCondition,
+}
+
+impl<'ctx> QueryCondition<'ctx> {
+    pub(crate) fn capi(&self) -> *mut ffi::tiledb_query_condition_t {
+        *self.raw
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn basic_op_test() -> TileDBResult<()> {
+        let qc1 = QC::field("field").lt(5);
+        let qc2 = QC::field("field").ge(6);
+        let qc = qc1 | qc2;
```

Comment:
IMO these tests should do a bit of basic pattern matching now, just to be sure that you actually got the right thing out.

Reply:
Totes fair. Will update.